tracardi/service/storage/drivers/elastic/tag.py

```diff
@@ -8,40 +8,42 @@ async def get_document_by_event_type(event_type):
 
 async def add_tags(event_type: str, tags: List[str]):
     search_result = await get_document_by_event_type(event_type=event_type)
-    if search_result.dict()["result"]:
-        tags.extend(search_result.dict()["result"][0]["tags"])
+    result_dict = search_result.dict()
```

Comment:
I do not know why you love dicts. Dicts are the worse possible solution for managing data. That is why driver is returning objects. If you are lacking some functionality in object then you can modify it. But most of the time you do not need that.

See How you can use object to do the same you did:

```
async def add_tags(event_type: str, tags: List[str]):
    search_result = await get_document_by_event_type(event_type=event_type)
    if search_result.total == 1:  # Now we are sure that there is only on record per event type
        record = list(search_result).pop()
        # TagForm should be moved from API to tracardi and named EventTag
        tag = EventTag(**record)  # now record is validated
        tag.tags.extend(tags)  # Appended tags
        tag.tags = list(set(tag.tags))  # Make it uniq
        return await storage_manager("event-tags").upsert(tag.dict())  # use object to save data
    else:
        return await storage_manager("event-tags").upsert({
            "type": event_type,
            "tags": tags
        })
```

Reply:
It is a lot of cleaner and is not so error prone as dicts.

Reply:
You should not convert SearchObjects to dict to process it.

Reply:
where exactly should TagForm be moved to? (I mean directory or file name)

Reply:
never mind, I've got it in second PR

Reply:
> where exactly should TagForm be moved to? (I mean directory or file name)

To tracardi repo, folder tracardi/domain